core/src/main/java/com/linecorp/armeria/client/retry/HttpStatusBasedRetryStrategy.java

```diff
@@ -37,28 +39,58 @@
  */
 final class HttpStatusBasedRetryStrategy implements RetryStrategy<HttpRequest, HttpResponse> {
 
-    private final List<HttpStatus> retryStatuses;
+    private Map<HttpStatus, Backoff> retryStatusesWithBackoff;
 
     /**
      * Creates a new instance.
      */
     HttpStatusBasedRetryStrategy(Iterable<HttpStatus> retryStatuses) {
-        this.retryStatuses = validateStatus(ImmutableList.copyOf(retryStatuses));
+        this(RetryStrategy.defaultBackoff, retryStatuses);
     }
 
-    private List<HttpStatus> validateStatus(List<HttpStatus> retryStatuses) {
-        requireNonNull(retryStatuses, "retryStatuses");
-        checkArgument(!retryStatuses.isEmpty(), "Need at least a status to retry");
+    /**
+     * Creates a new instance.
+     */
+    HttpStatusBasedRetryStrategy(Backoff backoff, Iterable<HttpStatus> retryStatuses) {
+        this(statusWithBackoff(requireNonNull(backoff, "backoff"),
+                               requireNonNull(retryStatuses, "retryStatuses")));
+    }
+
+    private static Map<HttpStatus, Backoff> statusWithBackoff(Backoff backoff,
+                                                              Iterable<HttpStatus> retryStatuses) {
+        checkArgument(retryStatuses.iterator().hasNext(), "need at least a status to retry");
+        final Builder<HttpStatus, Backoff> builder = ImmutableMap.builder();
+        for (HttpStatus httpStatus : retryStatuses) {
+            builder.put(httpStatus, backoff);
+        }
+        return builder.build();
+    }
+
+    /**
+     * Creates a new instance.
+     */
+    HttpStatusBasedRetryStrategy(Map<HttpStatus, Backoff> retryStatusesWithBackoff) {
+        validateRetryStatusesWithBackoff(retryStatusesWithBackoff);
+        this.retryStatusesWithBackoff =
+                retryStatusesWithBackoff instanceof ImmutableMap ? retryStatusesWithBackoff
+                                                                 : ImmutableMap.copyOf(
+                                                                         retryStatusesWithBackoff);
```

Comment:
What do you think about doing `ImmutableMap.copyOf()` before calling `validateRetryStatusesWithBackoff()` so that there's no chance for a user modifying the map during validation? (thread safety)

Please also note that `ImmutableMap.copyOf()` already handles the case where the specified map is an `ImmutableMap`.